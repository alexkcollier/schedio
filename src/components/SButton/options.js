export {
  colors,
  allShadeOptions,
  colorNames,
} from '@/utils/component-options'

export const sizes = [
  'small',
  'regular',
  'large',
]

export const types = [
  'outlined',
  'text',
]
