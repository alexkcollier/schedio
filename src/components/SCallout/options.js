export const calloutTypes = ['', 'success', 'warning', 'error']
