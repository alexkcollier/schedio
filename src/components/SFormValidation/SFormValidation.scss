.form-validation {
  display: flex;
  margin-top: spacing('half');

  &--success {
    color: color('green', 'dark');
  }

  &--warning {
    color: color('gold', 'dark');
  }

  &--error {
    color: color('red');
  }

  &__icon {
    flex: none;
    margin-right: spacing('half');
  }

  &__text {
    flex: auto;
  }
}
