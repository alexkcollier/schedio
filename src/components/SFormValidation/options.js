export const states = [
  '',
  'success',
  'warning',
  'error',
]
