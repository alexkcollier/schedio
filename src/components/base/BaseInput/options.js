export const allowed = ['email', 'text', 'password', 'url', 'search', ''].sort()
