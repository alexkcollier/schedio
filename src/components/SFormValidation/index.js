import SFormValidation from './SFormValidation.vue';

export { SFormValidation };

export default {
  install: Vue => Vue.component(SFormValidation.name, SFormValidation),
};
