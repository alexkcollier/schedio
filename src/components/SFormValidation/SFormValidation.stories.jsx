import { SFormValidation } from '@/components/SFormValidation'
import { SHeading } from '@/components/SHeading'
import PropList from '@@/docs/components/PropList'
import { boolean, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'
import { states } from './options'
import icons from 'feather-icons/dist/icons.json'
import { SInput } from '@/components/SInput'
import { SFormField } from '../SFormField/index'
import jeffsum from 'jeffsum'
import { generateHeading } from '@/utils/stories/render-functions'

storiesOf('Components/Forms.SFormValidation', module)
  .addParameters({ jest: 'SFormValidation' })
  .add('Form Validation', () => {
    return {
      props: {
        props: {
          default: () => ({
            text: text('text', 'validation text', 'Optional Props'),
            state: select('state', states, 'error', 'Optional Props'),
            icon: select('icon', ['', ...Object.keys(icons)], '', 'Optional Props'),
            hideIcon: boolean('hide-icon', false, 'Optional Props'),
          }),
        },
      },

      data () {
        return {
          inputOneValidity: null,
          shouldRespondInvalid: false,
          formValidity: null,
        }
      },

      methods: {
        handleSubmit (e) {
          e.preventDefault()
          this.formValidity = this.shouldRespondInvalid ? false : this.inputOneValidity
        },
      },
      render (h) {
        const { props } = this.$props

        return (
          <div>
            <SHeading level="1">Form Validation</SHeading>
            <p>
              The form validation component is to provides feedback on the validity of forms or
              their inputs. <code>{'<SFormValidation />'}</code> should be wrapped in the same{' '}
              <code>{'<SFormField />'}</code> as the element it is represeting.
            </p>

            <SHeading level="2">Standalone Example</SHeading>

            <SFormValidation {...{ props }} />

            <SHeading level="2">Usage Example</SHeading>

            <p>
              The input below will display validation messages on blur or keyup. The whole form will
              validate, based on the input and switch, when the button is pressed.
            </p>

            <form autocomplete="off" onsubmit={this.handleSubmit.bind(this)}>
              <SFormField>
                <SInput
                  id="email"
                  name="email"
                  label="Email address"
                  placeholder="email@domain.tld"
                  type="email"
                  onkeyup={(e) => { this.inputOneValidity = e.target.validity.valid }}
                  onblur={(e) => { this.inputOneValidity = e.target.validity.valid }}
                />

                <SFormValidation
                  v-show={this.inputOneValidity !== null}
                  state={this.inputOneValidity ? 'success' : 'error'}
                  text={this.inputOneValidity ? 'Valid!' : "That's not a valid email"}
                />
              </SFormField>

              <SFormField>
                <SSwitch
                  id="show-invalid"
                  name="show-invalid"
                  label="Respond as invalid"
                  value={this.shouldRespondInvalid}
                  oninput={(e) => { this.shouldRespondInvalid = e }}
                  required={false}
                  hideOptional={true}
                />
              </SFormField>

              <SFormField>
                <SButton button-color="green" disabled={!this.inputOneValidity}>
                  Validate the form
                </SButton>

                <SFormValidation
                  v-show={this.formValidity !== null}
                  state={this.formValidity ? 'success' : 'error'}
                  text={this.formValidity ? 'Valid!' : "The form's not valid"}
                />
              </SFormField>
            </form>

            {SFormValidation.props && <PropList component={SFormValidation} />}
          </div>
        )
      },
    }
  })
  .add(
    'Form Validation States',
    () => ({
      render (h) {
        return (
          <div>
            <SHeading>Form Validation States</SHeading>

            {states.map(state => generateValidations(h, state))}
          </div>
        )
      },
    }),
    {
      options: {
        showPanel: false,
      },
    }
  )

function generateValidations (h, state) {
  return [
    generateHeading(h, { level: 2, content: state || 'Base' }),
    h(SFormValidation, { props: { state, text: jeffsum(2) } }),
  ]
}
