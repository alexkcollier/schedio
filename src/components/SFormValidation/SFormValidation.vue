<template>
  <Transition name="fade">
    <div
      :class="classList"
      class="form-validation"
    >
      <div
        v-if="icons[state] && !hideIcon"
        class="form-validation__icon"
      >
        <SIcon
          :icon="icons[state]"
          :color="iconColors[state] || 'night'"
          aria-hidden="true"
        />
      </div>

      <div class="form-validation__text">
        <slot>{{ text }}</slot>
      </div>
    </div>
  </Transition>
</template>

<script>
import { SIcon } from '@/components/SIcon';
import { states } from './options';

export default {
  name: 'SFormValidation',

  components: {
    SIcon,
  },

  props: {
    text: {
      type: String,
      default: '',
    },

    state: {
      type: String,
      default: 'error',
      validator: (v) => {
        return (
          states.includes(v) || console.error(`\`state\` should be one of ${states.join(', ')}`)
        );
      },
    },

    icon: {
      type: String,
      default: '',
    },

    hideIcon: {
      type: Boolean,
      default: false,
    },
  },

  data () {
    return {
      iconColors: {
        success: 'green',
        warning: 'gold',
        error: 'red',
      },
    };
  },

  computed: {
    classList () {
      return {
        [`form-validation--${this.state}`]: this.state,
      };
    },

    icons () {
      if (this.icon) return { [this.state]: this.icon };

      return {
        success: 'check-circle',
        warning: 'alert-triangle',
        error: 'alert-circle',
      };
    },
  },
};
</script>

<style lang="scss">
.form-validation {
  display: flex;
  margin-top: spacing('half');

  &--success {
    color: color('green', 'dark');
  }

  &--warning {
    color: color('gold', 'dark');
  }

  &--error {
    color: color('red');
  }

  &__icon {
    flex: none;
    margin-right: spacing('half');
  }

  &__text {
    flex: auto;
  }
}
</style>
