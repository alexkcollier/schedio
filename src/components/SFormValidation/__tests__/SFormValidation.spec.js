import { shallowMount } from '@vue/test-utils'
import SFormValidation from '@/components/SFormValidation/SFormValidation.vue'
import SIcon from '@/components/SIcon/SIcon.vue'
import { states } from '../options'
import icons from 'feather-icons/dist/icons.json'
import { axe, toHaveNoViolations } from 'jest-axe'

expect.extend(toHaveNoViolations)

describe('SFormValidation.vue', () => {
  const errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => { })
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(SFormValidation)
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', async () => {
    expect(wrapper.html()).toMatchSnapshot()
    expect(await axe(wrapper.html())).toHaveNoViolations()
  })

  it('validates its visual state', () => {
    shallowMount(SFormValidation, { propsData: { state: 'not valid' } })

    expect(errorSpy).toHaveBeenCalled()
  })

  it('accepts text as a prop or slot', async () => {
    wrapper.setProps({ text: 'Some text' })
    await wrapper.vm.$nextTick()
    expect(wrapper.text()).toContain('Some text')
    expect(wrapper).toMatchSnapshot()

    const slotWrapper = shallowMount(SFormValidation, { slots: { default: 'More text' } })
    expect(slotWrapper.text()).toContain('More text')
    expect(slotWrapper).toMatchSnapshot()
  })

  states.forEach((state) => {
    it(`sets a visual state for ${state} state`, async () => {
      wrapper.setProps({ state })
      await wrapper.vm.$nextTick()
      expect(errorSpy).not.toHaveBeenCalled()
      expect(wrapper.html()).toMatchSnapshot()
      expect(await axe(wrapper.html())).toHaveNoViolations()
    })
  })

  it('accepts an icon prop', async () => {
    wrapper.setProps({ icon: Object.keys(icons)[0] })
    await wrapper.vm.$nextTick()
    expect(errorSpy).not.toHaveBeenCalled()

    const component = wrapper.findComponent(SIcon)
    expect(component.exists()).toBe(true)
    expect(wrapper).toMatchSnapshot()
  })

  it('can hide its icon', async () => {
    wrapper.setProps({ icon: Object.keys(icons)[0], hideIcon: true })
    await wrapper.vm.$nextTick()
    expect(errorSpy).not.toHaveBeenCalled()

    const component = wrapper.findComponent(SIcon)
    expect(component.exists()).toBe(false)
  })
})
