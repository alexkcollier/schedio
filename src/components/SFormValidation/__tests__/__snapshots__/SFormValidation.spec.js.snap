// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SFormValidation.vue accepts an icon prop 1`] = `
<transition-stub name="fade">
  <div class="form-validation form-validation--error">
    <div class="form-validation__icon">
      <sicon-stub icon="activity" color="red" shade="" size=""></sicon-stub>
    </div>
    <div class="form-validation__text"></div>
  </div>
</transition-stub>
`;

exports[`SFormValidation.vue accepts text as a prop or slot 1`] = `
<transition-stub name="fade">
  <div class="form-validation form-validation--error">
    <div class="form-validation__icon">
      <sicon-stub icon="alert-circle" color="red" shade="" size=""></sicon-stub>
    </div>
    <div class="form-validation__text">Some text</div>
  </div>
</transition-stub>
`;

exports[`SFormValidation.vue accepts text as a prop or slot 2`] = `
<transition-stub name="fade">
  <div class="form-validation form-validation--error">
    <div class="form-validation__icon">
      <sicon-stub icon="alert-circle" color="red" shade="" size=""></sicon-stub>
    </div>
    <div class="form-validation__text">More text</div>
  </div>
</transition-stub>
`;

exports[`SFormValidation.vue renders correctly 1`] = `
<transition-stub name="fade">
  <div class="form-validation form-validation--error">
    <div class="form-validation__icon">
      <sicon-stub icon="alert-circle" color="red" shade="" size=""></sicon-stub>
    </div>
    <div class="form-validation__text"></div>
  </div>
</transition-stub>
`;

exports[`SFormValidation.vue sets a visual state for  state 1`] = `
<transition-stub name="fade">
  <div class="form-validation">
    <!---->
    <div class="form-validation__text"></div>
  </div>
</transition-stub>
`;

exports[`SFormValidation.vue sets a visual state for error state 1`] = `
<transition-stub name="fade">
  <div class="form-validation form-validation--error">
    <div class="form-validation__icon">
      <sicon-stub icon="alert-circle" color="red" shade="" size=""></sicon-stub>
    </div>
    <div class="form-validation__text"></div>
  </div>
</transition-stub>
`;

exports[`SFormValidation.vue sets a visual state for success state 1`] = `
<transition-stub name="fade">
  <div class="form-validation form-validation--success">
    <div class="form-validation__icon">
      <sicon-stub icon="check-circle" color="green" shade="" size=""></sicon-stub>
    </div>
    <div class="form-validation__text"></div>
  </div>
</transition-stub>
`;

exports[`SFormValidation.vue sets a visual state for warning state 1`] = `
<transition-stub name="fade">
  <div class="form-validation form-validation--warning">
    <div class="form-validation__icon">
      <sicon-stub icon="alert-triangle" color="gold" shade="" size=""></sicon-stub>
    </div>
    <div class="form-validation__text"></div>
  </div>
</transition-stub>
`;
